Cypress open hangs for a minute or two before loading

### Current behavior
Hi ! 
For some time now, cypress open takes a couple minutes to start when running `npx cypress open`
Here are the steps we've encountered:

1.Cypress browser window is blank
![image](https://github.com/cypress-io/cypress/assets/14081633/3ff1558b-d76b-4de4-82e6-d366588ca4d0)

2. After a minute or so, the cypress loading logo appears and cypress works normally 
![image](https://github.com/cypress-io/cypress/assets/14081633/612c5a99-2c51-43cc-90f6-6ced4e2c4f01)

### Desired behavior

`npx cypress open` opens cypress window quickly

### Test code to reproduce
[bug-repro.zip](https://github.com/cypress-io/cypress/files/14693226/bug-repro.zip)



I dumped the output of `DEBUG=* npx cypress open --config-file cypress/cypress.local.config.ts  --browser edge` in this file : 
[out.log](https://github.com/cypress-io/cypress/files/14665230/out.log)


It seems that the app stops being stuck shortly after the following logs appear : 

```
2024-03-20T08:16:30.243Z cypress:data-context:sources:GitDataSource Refreshing git data
2024-03-20T08:16:30.243Z simple-git:task:rev-parse:4 Adding task to the queue, commands = [ 'rev-parse', '--show-toplevel' ]
2024-03-20T08:16:30.244Z simple-git:scheduler Scheduling id=4
2024-03-20T08:16:30.244Z simple-git:scheduler Attempting id=4
2024-03-20T08:16:30.244Z simple-git:task:rev-parse:4 Starting task
2024-03-20T08:16:30.244Z simple-git [GitExecutor] [SPAWN] git [ 'rev-parse', '--show-toplevel' ]
2024-03-20T08:16:30.244Z simple-git:task:rev-parse:4 [SPAWN] git [ 'rev-parse', '--show-toplevel' ]
2024-03-20T08:16:30.244Z simple-git:task:rev-parse:4 [SPAWN] {
  cwd: '/home/rozierale/Work/Rte/Naza/alz/naza-front',
  env: undefined,
  windowsHide: true
}
```
The "execution time" associated with the command `cypress:data-context:sources:GitDataSource Refreshing git data` was of 1minute (pretty long), so I dug a bit in cypress source code (`GitDataSource.ts`) , but it seems that the "1min" delay is due to this part : 
```
# GitDataSource.ts -> refreshAllGitData()
this.#intervalTimer = setTimeout(() => {
          this.#refreshAllGitData()
        }, SIXTY_SECONDS)
```
Note that **we don't have any delay with `cypress run`**, and this piece of code is ran only in open mode, that's why I thought this was relevant ! 
```
if (config.isRunMode) {
      this.#verifyGitRepo().catch(() => {
        // Empty catch for no-floating-promises rule
      })
    } else {
      this.#refreshAllGitData()
    }
```

**By downgrading to cypress 12.5.0, this issue disappears and cypress start normally. Here are the logs  of a successful execution** :
[working.log](https://github.com/cypress-io/cypress/files/14665347/working.log)

Interestingly, these logs do not contain the lines that appear when cypress stops being stuck (see above), so they must be unrelated to the issue....

And finally, cypress version 13.4.0  also has this issue (the only other I tried so far)

Sorry for the lack of other clues, I hope you'll have the time to take a look ! And thanks again for Cypress, we're using it everyday :)  

### Cypress Version

13.6.4

### Node version

v16.15.0

### Operating System

Ubuntu 20.04.5 LTS Release

### Debug Logs

```shell
see linked files.
```


### Other

_No response_

Hello,
I'm facing the same problem. Here is my two cents:

| Cypress versions  | Result |
| ------------- | ------------- |
| 12.5.0  | Launches quickly  |
| 12.6.0  | Launches quickly  |
| 13.0.0  | Launches quickly  |
| 13.1.0  | Launches quickly  |
| 13.2.0  | ~~Crashes with seemingly unrelated error about dependencies (Error: Cannot find module 'bluebird'), didn't investigate further~~ Hangs |
| 13.3.0  | Hangs  |

--- a/cli/CHANGELOG.md
+++ b/cli/CHANGELOG.md
@@ -11,6 +11,7 @@ _Released 5/7/2024 (PENDING)_
 
 - Fixed a bug where promises rejected with `undefined` were failing inside `cy.origin()`. Addresses [#23937](https://github.com/cypress-io/cypress/issues/23937).
 - We now pass the same default Chromium flags to Electron as we do to Chrome. As a result of this change, the application under test's `navigator.webdriver` property will now correctly be `true` when testing in Electron. Fixes [#27939](https://github.com/cypress-io/cypress/issues/27939).
+- Fixed network issues in requests using fetch for users where Cypress is run behind a proxy that performs HTTPS decryption (common among corporate proxies). Fixes [#29171](https://github.com/cypress-io/cypress/issues/29171)
 
 **Misc:**
 

--- a/packages/network/test/support/servers.ts
+++ b/packages/network/test/support/servers.ts
@@ -12,14 +12,18 @@ export interface AsyncServer {
   listenAsync: (port) => Promise<void>
 }
 
-function createExpressApp () {
+function createExpressApp (requestCallback: (req) => void) {
   const app: express.Application = express()
 
   app.get('/get', (req, res) => {
+    if (requestCallback) requestCallback(req)
+
     res.send('It worked!')
   })
 
   app.get('/empty-response', (req, res) => {
+    if (requestCallback) requestCallback(req)
+
     // ERR_EMPTY_RESPONSE in Chrome
     setTimeout(() => res.connection.destroy(), 100)
   })
@@ -45,10 +49,11 @@ export class Servers {
   wsServer: any
   wssServer: any
   caCertificatePath: string
+  lastRequestHeaders: any
 
   start (httpPort: number, httpsPort: number) {
     return Promise.join(
-      createExpressApp(),
+      createExpressApp((req) => this.lastRequestHeaders = req.headers),
       getCAInformation(),
     )
     .spread((app: Express.Application, [serverCertificateKeys, caCertificatePath]: [serverCertificateKeys: string[], caCertificatePath: string]) => {

--- a/packages/network/test/unit/agent_spec.ts
+++ b/packages/network/test/unit/agent_spec.ts
@@ -227,6 +227,15 @@ describe('lib/agent', function () {
           })
         })
 
+        it('HTTP pages are requested with correct host header when loaded via fetch', function () {
+          return this.fetch(`http://localhost:${HTTP_PORT}/get`)
+          .then(() => {
+            expect(this.servers.lastRequestHeaders).to.include({
+              host: `localhost:${HTTP_PORT}`,
+            })
+          })
+        })
+
         it('HTTPS pages can be loaded', function () {
           return this.request({
             url: `https://localhost:${HTTPS_PORT}/get`,
@@ -255,6 +264,15 @@ describe('lib/agent', function () {
           })
         })
 
+        it('HTTPS pages are requested with correct host header when loaded via fetch', function () {
+          return this.fetch(`https://localhost:${HTTPS_PORT}/get`)
+          .then(() => {
+            expect(this.servers.lastRequestHeaders).to.include({
+              host: 'localhost',
+            })
+          })
+        })
+
         it('HTTPS pages can be loaded via fetch with no explicit port', function () {
           return this.fetch(`https://localhost/get`)
           .then((response) => response.text())
@@ -266,6 +284,15 @@ describe('lib/agent', function () {
                 url: `localhost:${HTTPS_PORT}`,
               })
             }
+          })
+        })
+
+        it('HTTPS pages requested with correct host header when loaded via fetch with no explicit port', function () {
+          return this.fetch(`https://localhost/get`)
+          .then(() => {
+            expect(this.servers.lastRequestHeaders).to.include({
+              host: 'localhost',
+            })
           })
         })
 

--- a/patches/node-fetch+2.7.0.patch
+++ b/patches/node-fetch+2.7.0.patch
@@ -8,7 +8,7 @@ index 567ff5d..c7e2bd9 100644
  
 -		const send = (options.protocol === 'https:' ? https : http).request;
 +		const isHttps = options.protocol === 'https:';
-+		options.defaultPort = opts.defaultPort || (isHttps ? 443 : 80);
++		options.defaultPort = (isHttps ? 443 : 80);
 +		const send = (isHttps ? https : http).request;
  		const signal = request.signal;
  

--- a/patches/node-fetch+2.7.0.patch
+++ b/patches/node-fetch+2.7.0.patch
@@ -1,0 +1,15 @@
+diff --git a/node_modules/node-fetch/lib/index.js b/node_modules/node-fetch/lib/index.js
+index 567ff5d..c7e2bd9 100644
+--- a/node_modules/node-fetch/lib/index.js
++++ b/node_modules/node-fetch/lib/index.js
+@@ -1449,7 +1449,9 @@ function fetch(url, opts) {
+ 		const request = new Request(url, opts);
+ 		const options = getNodeRequestOptions(request);
+ 
+-		const send = (options.protocol === 'https:' ? https : http).request;
++		const isHttps = options.protocol === 'https:';
++		options.defaultPort = opts.defaultPort || (isHttps ? 443 : 80);
++		const send = (isHttps ? https : http).request;
+ 		const signal = request.signal;
+ 
+ 		let response = null;

